test/resource_owner_password_grant_test.dart

```diff
@@ -0,0 +1,109 @@
+  "error": "unauthorized",
+  "error_description": "unauthorized",
+  "error_uri": "http://error.com"
+}));
+
+final List<int> SUCCESS = UTF8.encode(JSON.encode({
+  "access_token": "2YotnFZFEjr1zCsicMWpAA",
+  "token_type": "bearer",
+  "expires_in": 3600,
+  "refresh_token": "tGzv3JOkF0XG5Qx2TlKWIA",
+}));
+
+void main() {
+  group('basic', () {
+    HttpServer server;
+    var authorizationEndpoint =
+        Uri.parse('http://${InternetAddress.LOOPBACK_IP_V4.host}:8080');
+
+    setUp(() async {
+      server = await HttpServer.bind(InternetAddress.LOOPBACK_IP_V4, 8080);
+    });
+
+    test('builds correct request with client when using basic auth for client',
+        () async {
+      server.take(1).listen((HttpRequest request) {
+        request.response.headers.add(
+            HttpHeaders.CONTENT_TYPE, 'application/json;charset=UTF-8');
+        var auth =
+            'Basic ${CryptoUtils.bytesToBase64(UTF8.encode('client:secret'))}';
+
+        expect(auth, equals(request.headers['authorization'].first));
+        expect(request.uri.queryParameters['grant_type'], equals('password'));
+        expect(request.uri.queryParameters['username'], equals('username'));
+        expect(request.uri.queryParameters['password'], equals('userpass'));
+        request.response.statusCode = HttpStatus.OK;
+        request.response.add(SUCCESS);
+        request.response.close();
+      });
+
+      var client = await oauth2.resourceOwnerPasswordGrant(
+          authorizationEndpoint, 'username', 'userpass',
+          clientId: 'client', clientSecret: 'secret');
+      expect(client.credentials, isNotNull);
+      expect(client.credentials.accessToken, equals('2YotnFZFEjr1zCsicMWpAA'));
+    }, timeout: new Timeout(new Duration(seconds: 1)));
+
+    test('builds correct request when using query parameters for client',
+        () async {
+      server.take(1).listen((HttpRequest request) {
+        request.response.headers.add(
+            HttpHeaders.CONTENT_TYPE, 'application/json;charset=UTF-8');
+        expect(request.uri.queryParameters['grant_type'], equals('password'));
+        expect(request.uri.queryParameters['client_id'], equals('client'));
+        expect(request.uri.queryParameters['client_secret'], equals('secret'));
+        expect(request.uri.queryParameters['username'], equals('username'));
+        expect(request.uri.queryParameters['password'], equals('userpass'));
+        request.response.statusCode = HttpStatus.OK;
+        request.response.add(SUCCESS);
+        request.response.close();
+      });
+
+      var client = await oauth2.resourceOwnerPasswordGrant(
+          authorizationEndpoint, 'username', 'userpass',
+          clientId: 'client', clientSecret: 'secret', useBasicAuth: false);
+      expect(client.credentials, isNotNull);
+      expect(client.credentials.accessToken, equals('2YotnFZFEjr1zCsicMWpAA'));
+    }, timeout: new Timeout(new Duration(seconds: 1)));
```

Comment:
Also test that the query parameters are properly merged with existing query parameters.